src/libstd/time/iso_8601.rs

```diff
@@ -0,0 +1,331 @@
+        cycle_start += cycles as i64 * 400;
+        let day_of_cycle = days % CYCLE_DAYS as u64;
+
+        let (year_mod_400, day_of_year) =
+            TmUtc::day_of_cycle_to_year_day_of_year(day_of_cycle as u32);
+
+        let (year,) = (cycle_start + year_mod_400,);
+        let (month, day) = TmUtc::day_of_year_to_month_day(year, day_of_year);
+        let (hour, minute, second) =
+            TmUtc::seconds_of_day_to_h_m_s(duration_of_day.as_secs() as u32);
+
+        TmUtc {
+            year,
+            month,
+            day,
+            hour,
+            minute,
+            second,
+            nanos: duration_of_day.subsec_nanos(),
+        }
+    }
+
+    // Construct from duration added to epoch
+    pub fn from_epoch_add(add: Duration) -> TmUtc {
+        // Special case to prevent integer overflow in duration addition
+        if add > Duration::from_secs(YEARS_1970_2000_SECONDS) {
+            TmUtc::from_cycle_start_add_duration(
+                2000, add - Duration::from_secs(YEARS_1970_2000_SECONDS))
+        } else {
+            TmUtc::from_cycle_start_add_duration(
+                1600, add + Duration::from_secs(YEARS_1600_1970_SECONDS))
+        }
+    }
+
+    // Construct from duration subtracted from epoch
+    pub fn from_epoch_sub(mut sub: Duration) -> TmUtc {
+
+        // Make `sub` less than 400 years
+
+        // Number of full leap cycles in `sub`
+        let leap_cycles = sub.as_secs() / CYCLE_SECONDS;
+        let year = 1970 - (leap_cycles * 400) as i64;
+        sub -= Duration::from_secs(leap_cycles * CYCLE_SECONDS);
+
+        // Convert subtraction to addition
+        // and align year to a cycle start
+
+        // `year` is still 370 years ahead of cycle start.
+        debug_assert!((year % 400 + 400) % 400 == 370);
+
+        // Subtract 370 years to get to cycle start
+        // and additional 400 year to prevent integer underflow in duration subtraction
+        let cycle_start = year - 400 - 370;
+        let add_from_cycle_start =
+            Duration::from_secs(CYCLE_SECONDS + YEARS_1600_1970_SECONDS) - sub;
+
+        debug_assert!(cycle_start % 400 == 0);
+
+        return TmUtc::from_cycle_start_add_duration(cycle_start, add_from_cycle_start);
+    }
+
+    pub fn fmt_iso_8601(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.year > 9999 {
+            // ISO-8601 extended format requires leading plus
+            write!(f, "+{}", self.year)?;
+        } else if self.year < 0 {
+            // At least 4 digits
+            write!(f, "{:05}", self.year)?;
+        } else {
+            write!(f, "{:04}", self.year)?;
+        }
+
+        write!(f, "-{:02}-{:02}T{:02}:{:02}:{:02}",
+            self.month, self.day, self.hour, self.minute, self.second)?;
+
+        // if width is not specified, print nanoseconds
+        let subsec_digits = f.precision().unwrap_or(9);
+        if subsec_digits != 0 {
+            let mut subsec_digits = subsec_digits;
```

Comment:
move this line down to just before `subsec_digits` is mutated? does that even make sense?